engine/lib/actions.php

```diff
@@ -98,6 +98,38 @@ function elgg_unregister_action($action) {
 }
 
 /**
+ * Generate an authentication code/token in Base64URL encoding, using the site secret as key.
+ *
+ * As a MAC often needs to validate several values, you can pass an array of strings as $data.
+ *
+ * @param string[]|string $data Data passed through an untrusted channel
+ * @param string          $algo Hash algorithm used in HMAC
+ *
+ * @return string
+ * @since 1.11
+ * @see elgg_validate_mac
+ */
+function elgg_generate_mac($data, $algo = 'sha256') {
+	return _elgg_services()->crypto->getHmac($data, '', $algo);
+}
+
+/**
+ * Validate an authentication code generated via elgg_generate_mac().
+ *
+ * @param string          $mac  MAC given in input, in Base64URL format
+ * @param string[]|string $data Data ostensibly unaltered by the user
+ * @param string          $algo Hash algorithm used in HMAC
+ *
+ * @return bool
+ * @since 1.10
+ * @see elgg_generate_mac
+ */
+function elgg_validate_mac($mac, $data, $algo = 'sha256') {
+	$crypto = _elgg_services()->crypto;
+	return $crypto->areEqual($mac, $crypto->getHmac($data, '', $algo));
+}
+
+/**
```

Comment:
do these functions really need to be in actions? I don't know a really good location either